translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,377 @@
+// limitations under the License.
+
+package metrics
+
+import (
+	"sort"
+
+	ocmetrics "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	otlpcommon "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1"
+	otlpmetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	translatorcommon "github.com/open-telemetry/opentelemetry-collector/translator/common"
+)
+
+const (
+	invalidMetricDescriptorType = ocmetrics.MetricDescriptor_Type(-1)
+)
+
+// ResourceMetricsToMetricsData converts metrics from OTLP to internal (OpenCensus) format
+func ResourceMetricsToMetricsData(resourceMetrics *otlpmetrics.ResourceMetrics) consumerdata.MetricsData {
+
+	node, resource := translatorcommon.ResourceToOC(resourceMetrics.Resource)
+	md := consumerdata.MetricsData{
+		Node:     node,
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	labelKeys := allLabelKeys(metric)
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor, labelKeys),
+		Timeseries:       dataPointsToTimeseries(metric, labelKeys),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor, labelKeys []*ocmetrics.LabelKey) *ocmetrics.MetricDescriptor {
+	if descriptor == nil {
+		return nil
+	}
+
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        descriptorTypeToOC(descriptor.Type),
+		LabelKeys:   labelKeys,
+	}
+}
+
+func allLabelKeys(metric *otlpmetrics.Metric) []*ocmetrics.LabelKey {
+	// NOTE: OpenTelemetry and OpenCensus have different representations of labels:
+	// - OC has a single "global" ordered list of label keys per metric in the MetricDescriptor;
+	// then, every data point has an ordered list of label values matching the key index.
+	// - In OTLP, every label is represented independently via StringKeyValue,
+	// i.e. theoretically points in the same metric may have different set of labels.
+	//
+	// So what we do in this translator:
+	// - Scan all points and their labels to generate a unique set of all label keys
+	// used across the metric, sort them and set in MetricDescriptor.
+	// - For each point we generate an ordered list of label values,
+	// matching the order of label keys returned here (see `labelsToOC` function).
```

Comment:
Probably worth mentioning that some values for some generated data points may be listed as "unspecified" since the corresponding values are not present in the source data points.

Reply:
Done.